File: python/behavioral/memento.py
#!/usr/bin/python
# -*- coding : utf-8 -*-

"""
brief:
  Memento.
    Without violating encapsulation,capture
  and externalize an object's internal state so that the
  object can be restored to this state later.
  
          Originator         <==>      Caretaker
              |                           |
   Create and restore Memento    Save and provide Memento
"""

# Memento. Responsible for storing the internal state of the Originator
#       and providing the internal state of the Originator as needed.
class Memento: 
  def __init__(self, state):
    self.state = state
    
  def set_state(self, state):
    self.state = state
    
  def get_state(self):
    return self.state
    
# Originator. It is responsible for recording the internal state of the
#        current moment, defining which states belong to the backup scope,
#        and creating and restoring the memo data.
class Originator:
  def __init__(self, state):
    self.state = state
    
  def create_memento(self):
    return Memento(self.state)

  def restore_from_memento(self, memento):
    self.state = memento.get_state();
    
  def set_state(self, state):
    self.state = state
    
  def get_state(self):
    return self.state  
    
# Caretaker. Manage, save and provide the Memento.
class Caretaker:  
  def set_memento(self, memento):
    self.memento = memento

  def get_memento(self):
    return self.memento
  
if __name__ == '__main__':

  # Initialize.
  org = Originator("A")
  print("%s" %(org.get_state()))
  
  # Save.
  taker = Caretaker();
  taker.set_memento(org.create_memento());
  
  # Change.
  org.set_state("B")
  print("%s" %(org.get_state()))
  
  # Restore.
  org.restore_from_memento(taker.get_memento());
  print("%s" %(org.get_state()))
  

File: design-patterns/cpp/behavioral/memento.cpp
/*!
* \brief   Memento. Without violating encapsulation,capture
*       and externalize an object's internal state so that the
*       object can be restored to this state later.
*
*/
#include <iostream>

// Memento. Responsible for storing the internal state of the Originator
//       and providing the internal state of the Originator as needed.
class Memento {
public:
  Memento(const std::string& state) {
    this->state_ = state;
  }
  void SetState(const std::string& state) {
    this->state_ = state;
  }
  std::string &GetState() {
    return this->state_;
  }
private:
  std::string state_;
};

// Originator. It is responsible for recording the internal state of the
//        current moment, defining which states belong to the backup scope,
//        and creating and restoring the memo data.
class Originator {
public:
  Originator(const std::string& state) {
    this->state_ = state;
  }
  Memento* CreateMemento() {
    return new Memento(this->state_);
  }
  void RestoreToMemento(Memento* memento) {
    this->state_ = memento->GetState();
  }
  void SetState(const std::string& state) {
    this->state_ = state;
  }
  std::string &GetState() {
    return this->state_;
  }

private:
  std::string state_;
};

// Caretaker. Manage, save and provide the Memento.
class Caretaker {
public:
  void SetMemento(Memento* memento) {
    this->memento_ = memento;
  }
  Memento* GetMemento() {
    return this->memento_;
  }

private:
  Memento* memento_;
};

int main() {
  // Initialize.
  Originator* org = new Originator("A");
  std::cout << org->GetState().c_str() << std::endl;

  // Save.
  Caretaker* taker = new Caretaker();
  taker->SetMemento(org->CreateMemento());

  // Change.
  org->SetState("B");
  std::cout << org->GetState().c_str() << std::endl;

  // Restore.
  org->RestoreToMemento(taker->GetMemento());
  std::cout << org->GetState().c_str() << std::endl;

  return 0;
}